wdl-lint/src/rules/expression_spacing.rs

```diff
@@ -0,0 +1,726 @@
+         of your expression and it needs to be line split, the entire construct should be wrapped \
+         in parentheses (`()`). The opening parenthesis should be immediately followed by a \
+         newline. `if`, `then`, and `else` should all start a line one more level of indentation \
+         than the wrapping paratheses. The closing parenthesis should be on the same level of \
+         indentation as the opening parenthesis. If you are using the `if...then...else...` \
+         construct on one line, it does not need to be wrapped in parentheses. However, if any of \
+         the 3 clauses are more complex than a single identifier, they should be wrapped in \
+         parentheses. Sometimes a developer will choose to line split an expression despite it \
+         being able to all fit on one line that is <=90 characters wide. That is perfectly \
+         acceptable, though you may notice in the below examples the single line form can be more \
+         readable. There is 'wiggle' room allowed by the above rules. This is intentional, and \
+         allows developers to choose a more compact or a more spaced out expression."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Spacing])
+    }
+}
+
+impl Visitor for ExpressionSpacingRule {
+    type State = Diagnostics;
+
+    fn document(
+        &mut self,
+        _: &mut Self::State,
+        reason: VisitReason,
+        _: &Document,
+        _: SupportedVersion,
+    ) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        // Reset the visitor upon document entry
+        *self = Default::default();
+    }
+
+    fn expr(&mut self, state: &mut Self::State, reason: VisitReason, expr: &Expr) {
+        if reason == VisitReason::Exit {
+            return;
+        }
+
+        match expr {
+            Expr::LogicalNot(_) | Expr::Negation(_) => {
+                // No following spacing allowed
+                if expr
+                    .syntax()
+                    .children_with_tokens()
+                    .filter(|t| t.kind() == SyntaxKind::Whitespace)
+                    .count()
+                    > 0
+                {
+                    state.add(prefix_whitespace(expr.syntax().text_range().to_span()));
+                }
+            }
+            Expr::Parenthesized(_) => {
+                // Find the actual open and close parentheses
+                let open = expr
+                    .syntax()
+                    .children_with_tokens()
+                    .find(|t| t.kind() == SyntaxKind::OpenParen)
+                    .expect("parenthesized expression should have an opening parenthesis");
+                let close = expr
+                    .syntax()
+                    .children_with_tokens()
+                    .find(|t| t.kind() == SyntaxKind::CloseParen)
+                    .expect("parenthesized expression should have an closing parenthesis");
```

Comment:
I'm not sure this logic is 100% correct. `find()` returns the first matching token. E.g. `Int expr = ((x+42))` I think it might return the inner `CloseParen` when you want it to return the outer. I guess it depends on what's included as a "child" of the `Expr`? Maybe your logic works, I'm not sure.

Reply:
In your example, the first `CloseParen` would be a child of the inner `ParenthesizedExpression`.

Reply:
So your logic here should be ok then. There can only ever be 1 `CloseParen` child, any additional `CloseParen` will be nested under something else.